AzureMonitor: User authentication support

aangelisc opened the pull request:
This PR adds support for current user authentication to the Azure Monitor data source. It's quite a large PR but its been done in segments with each PR reviewed individually.

`grafana` PR's:
- #82332 
- #83107
- #83440 

`grafana-azure-sdk-go` PR's:
- grafana/grafana-azure-sdk-go#68
- grafana/grafana-azure-sdk-go#99
- grafana/grafana-azure-sdk-go#111

In order to support user-based authentication in Azure Monitor the following has been done:

- Bump `grafana-azure-sdk-go` to v1.13.1. This adds support for current user authentication fallback credentials and adds support for these credentials in the credentials builder.
- Add `user_identity_fallback_credentials_enabled` to the Grafana configuration. This defaults to `true` when `user_identity_enabled` is also `true`. This configuration variable allows Grafana administrators to enable/disable fallback credentials at the instance level. 
    - The `sample.ini` file has been updated.
    - The docs have been updated.
    - The frontend types have been updated.
    - The environment variable configuration has been updated.
- Azure Monitor `QueryData`, `CallResource`, and `CheckHealth` functions have been updated to appropriately pass the user context.
- Azure Monitor now allows user identity authentication.
- `ConfigEditor` components have been co-located, similar to the various `QueryEditor` components.
- Credentials logic has been updated for current user authentication.
- Updates the frontend to better support user-based authentication.
- The config editor has been updated to allow configuration of fallback service credentials (App Registration, Managed Identity, and Workload Identity are currently supported).
- Add toggle for enabling/disabling fallback credentials at the data source level (in addition to the instance level toggle).
- Add alert if fallback credentials are disabled at the instance level.
- Makes use of the new style credentials given that user-auth hasn't existed in Azure Monitor previously.
- Adds various warnings around user-auth and what may/may not work.
- Add `authenticatedBy` prop to user object in `@grafana/runtime`, making it available to data sources.
- Update mocks
- Update some tests
- Update docs

This will benefit from manual user testing. Let me know if you need help to achieve this.

Fallback credentials disabled alert:
![image](https://github.com/grafana/grafana/assets/15019026/29760ff7-300b-4d58-848f-2bfe8147720c)

Fallback credentials information alert:
![image](https://github.com/grafana/grafana/assets/15019026/0ba5e64c-0f3f-4b35-9b32-b93bd3b1eed8)

Fallback credentials configuration:
![image](https://github.com/grafana/grafana/assets/15019026/c44349aa-1206-4e3a-98f4-20bf14f80bb4)

Alerting warning when fallback credentials are missing/disabled:
![image](https://github.com/grafana/grafana/assets/15019026/ec7a075e-daa7-4241-b539-76e96749beec)

Query editor warning when logged in with auth provider other than Azure:
![image](https://github.com/grafana/grafana/assets/15019026/d989e663-c6ef-47d8-9b83-f65e8d9a2edb)

Part of #81918 

Note: This feature is experimental and may be subject to unexpected behaviour or potential breaking changes. Backend Grafana functionality e.g. alerts, recorded queries, reporting etc may not function as expected with current user authentication.

Fixes #85635

aangelisc committed:
Stub out frontend user auth

aangelisc committed:
Stub out backend user auth

aangelisc committed:
Add context

aangelisc committed:
Reorganise files

aangelisc committed:
Refactor app registration form

aangelisc committed:
Alert for user auth service principal credentials

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	pkg/tsdb/azuremonitor/httpclient.go

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	public/app/plugins/datasource/azuremonitor/components/AzureCredentialsForm.tsx

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	pkg/tsdb/azuremonitor/azuremonitor.go
#	public/app/plugins/datasource/azuremonitor/components/QueryEditor/QueryEditor.tsx
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/AdvancedMulti.tsx
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/NestedEntry.tsx
#	public/app/plugins/datasource/azuremonitor/components/ResourcePicker/ResourcePicker.tsx
#	public/app/plugins/datasource/azuremonitor/components/VariableEditor/VariableEditor.tsx
#	public/app/plugins/datasource/azuremonitor/components/shared/Space.tsx

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
AzureMonitor: Add flag for enabling/disabling fallback credentials for current user authentication (#82332)

* Rename field

* Add fallback setting

* Update tests and mock

* Remove duplicate setting line

* Update name of property

* Update frontend settings

* Update docs and default config files

* Update azure-sdk

* Fix lint

* Update test

* Bump dependency

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	pkg/plugins/envvars/envvars.go
#	pkg/services/pluginsintegration/pluginconfig/envvars_test.go

aangelisc committed:
Update configuration

aangelisc commented:
Mistakenly marked this as ready for review too early, sorry for the spam

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -1203,6 +1203,12 @@ Specifies whether user identity authentication (on behalf of currently signed-in
 
+### user_identity_fallback_credentials_enabled
+
+Specifies whether user identity authentication fallback credentials should be enabled in data sources. Enabling this allows data source creators to provide fallback credentials for backend initiated requests e.g. alerting, recorded queries etc.
```
```suggestion
Specifies whether user identity authentication fallback credentials should be enabled in data sources. Enabling this allows data source creators to provide fallback credentials for backend-initiated requests, such as alerting, recorded queries, and so on.
```

chri2547 commented on docs/sources/setup-grafana/configure-grafana/_index.md:
```diff
@@ -1203,6 +1203,12 @@ Specifies whether user identity authentication (on behalf of currently signed-in
+
+Specifies whether user identity authentication fallback credentials should be enabled in data sources. Enabling this allows data source creators to provide fallback credentials for backend initiated requests e.g. alerting, recorded queries etc.
+
+Enabled by default, needs to be explicitly disabled. Will not have any effect if user identity is disabled above.
```
```suggestion
It is enabled by default and needs to be explicitly disabled. It will not have any effect if the user identity is disabled.
```

chri2547 reviewed (commented):
A few docs suggestions. Please take a look to ensure I didn't change the technical meaning. Thank you for the contribution!

aangelisc committed:
Update docs/sources/setup-grafana/configure-grafana/_index.md

Co-authored-by: Christopher Moyer <35463610+chri2547@users.noreply.github.com>

aangelisc committed:
Docs review

chri2547 reviewed (commented):
Docs approved! Thanks for the contribution!

aangelisc committed:
AzureMonitor: User authentication frontend updates (#83107)

* Rename field

* Add fallback setting

* Update tests and mock

* Remove duplicate setting line

* Update name of property

* Update frontend settings

* Update docs and default config files

* Add alerts to query editor

- Add authenticatedBy property to grafana/data
- Update mocks
- Update query editor to disable it under certain circumstances
- Update tests

* Add separate FallbackCredentials component

- Reset AppRegistrationCredentials component to only handle clientsecret credentials
- Update AzureCredentialsForm
- Update selectors
- Update tests
- Update credentials utility functions logic

* Alert when fallback credentials disabled

* Update condition

* Update azure-sdk

* Fix lint

* Update test

* Remove unneeded conditions

* Set auth type correctly

* Legacy cloud options

* Fix client secret

* Remove accidental import

* Bump dependency

* Add tests

* Don't use VerticalGroup component

* Remove unused import

* Fix lint

* Appropriately set oAuthPassThru and disableGrafanaCache properties

* Clear azureCredentials on authType change

* Correctly retrieve secret

* Fix bug in authTypeOptions

* Update public/app/plugins/datasource/azuremonitor/components/ConfigEditor/CurrentUserFallbackCredentials.tsx

Co-authored-by: Andrew Hackmann <5140848+bossinc@users.noreply.github.com>

* Update public/app/plugins/datasource/azuremonitor/components/QueryEditor/QueryEditor.tsx

Co-authored-by: Andrew Hackmann <5140848+bossinc@users.noreply.github.com>

* Update public/app/plugins/datasource/azuremonitor/components/ConfigEditor/CurrentUserFallbackCredentials.tsx

Co-authored-by: Andrew Hackmann <5140848+bossinc@users.noreply.github.com>

* Add documentation links

* Fix broken link

---------

Co-authored-by: Andrew Hackmann <5140848+bossinc@users.noreply.github.com>

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	go.mod

aangelisc committed:
AzureMonitor: Update docs for current user authentication (#83440)

* Rename field

* Add fallback setting

* Update tests and mock

* Remove duplicate setting line

* Update name of property

* Update frontend settings

* Update docs and default config files

* Add alerts to query editor

- Add authenticatedBy property to grafana/data
- Update mocks
- Update query editor to disable it under certain circumstances
- Update tests

* Add separate FallbackCredentials component

- Reset AppRegistrationCredentials component to only handle clientsecret credentials
- Update AzureCredentialsForm
- Update selectors
- Update tests
- Update credentials utility functions logic

* Alert when fallback credentials disabled

* Update condition

* Update azure-sdk

* Fix lint

* Update test

* Remove unneeded conditions

* Set auth type correctly

* Legacy cloud options

* Fix client secret

* Remove accidental import

* Bump dependency

* Add tests

* Don't use VerticalGroup component

* Remove unused import

* Update docs

* Fix lint

* Appropriately set oAuthPassThru and disableGrafanaCache properties

* Clear azureCredentials on authType change

* Correctly retrieve secret

* Feedback

* Spelling

* Update docs/sources/datasources/azure-monitor/_index.md

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

* Update docs/sources/datasources/azure-monitor/_index.md

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

* Update docs/sources/datasources/azure-monitor/_index.md

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

* Update docs/sources/datasources/azure-monitor/_index.md

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

---------

Co-authored-by: Larissa Wandzura <126723338+lwandz13@users.noreply.github.com>

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	go.mod

jdbaldry reviewed (approved):
Some more style suggestions for the documentation changes.

Let me know if I can explain any in more detail :)

aangelisc commented:
/deploy-to-hg

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

aangelisc commented on docs/sources/datasources/azure-monitor/_index.md:
```diff
@@ -201,6 +226,39 @@ For details on workload identity, refer to the [Azure workload identity document
 
+### Configure Current User Authentication
+
+{{< admonition type="note" >}}
+This feature is experimental and subject to change or removal. Aspects of Grafana may not work as expected when using this authentication method.
+{{< /admonition >}}
```
As discussed have written a slightly modified version of this.

aangelisc commented on docs/sources/datasources/azure-monitor/_index.md:
```diff
@@ -201,6 +226,39 @@ For details on workload identity, refer to the [Azure workload identity document
+To configure fallback service credentials the [feature toggle][configure-grafana-feature-toggles] `idForwarding` must be set to `true` and `user_identity_fallback_credentials_enabled` must be enabled in the [Azure configuration section][configure-grafana-azure] (enabled by default when `user_identity_enabled` is set to `true`).
+{{< /admonition >}}
+
+Permissions for fallback credentials may need to be broad in order to appropriately support backend functionality. For example, an alerting query created by a user will be dependent on their permissions. If a user creates an alert for a resource that the fallback credentials cannot access, the alert will fail.
```
Have made this change but I've kept `the alert will fail` rather than `the creation fails` as the creation will succeed but the alert itself will error.

aangelisc commented on docs/sources/datasources/azure-monitor/_index.md:
```diff
@@ -201,6 +226,39 @@ For details on workload identity, refer to the [Azure workload identity document
+
+{{< figure src="/media/docs/grafana/data-sources/screenshot-current-user.png" max-width="800px" class="docs-image--no-shadow" caption="Azure Monitor screenshot showing Current User authentication" >}}
+
+3. If backend functionality is desired using this data source, enable service credentials and configure the data source using the most applicable credentials for your circumstances.
```
Have made this change but dropped the ``` at the end of the line.

aangelisc committed:
Docs review

aangelisc committed:
Update docs with additional configuration information

aangelisc committed:
Fix to appropriately hide the query editor

aangelisc committed:
Typo

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

jdbaldry reviewed (approved):
LGTM! Thank you for considering the suggestions and writing these docs in the first place :)

bossinc reviewed (approved):
This was a lot of work! Very nice test coverage. Thank you for splitting up the PRs by PRing into this branch
:tada: :tada: :tada:

I gave it a once over, and nothing stuck out. I didn't test alerting and recorded queries. Let me know if I should tests these

aangelisc commented:
/deploy-to-hg

aangelisc committed:
Update isCredentialsComplete

aangelisc commented:
/deploy-to-hg

aangelisc committed:
Merge branch 'main' into andreas/az-user-auth

# Conflicts:
#	go.mod
#	go.sum
#	pkg/services/pluginsintegration/pluginconfig/envvars_test.go
#	pkg/tsdb/azuremonitor/azuremonitor.go

aangelisc committed:
Update test
